### PR TITLE
WT-7375 Implement minimal drop support for tiered storage.

### PR DESCRIPTION
**Pull request by sueloverso**

### CODE REVIEW

File: src/schema/schema_drop.c
Comment (keitharnoldsmith) [inline src/schema/schema_drop.c:201]: Does this need a matching call to `wt_meta_track_drop` to ensure that the local file gets removed?

File: src/schema/schema_drop.c
Comment (sueloverso) [reply src/schema/schema_drop.c:208]: Fixed.

File: src/schema/schema_drop.c
Comment (sueloverso) [reply src/schema/schema_drop.c:201]: Good catch. I also, therefore, added support for the `remove_files` configuration setting. And removed the `wt_meta_track_drop` from the object loop because we're not removing objects.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:564]: Not quite the case I had in mind, but yes, it covers that too! I think that does not belong in the comment however.

File: test/suite/test_tiered07.py
Comment (sueloverso) [reply test/suite/test_tiered07.py:80]: Fixed.

File: src/tiered/tiered_handle.c
Comment (ddanderson) [inline src/tiered/tiered_handle.c:32]: Do we have to prune the list that we get from `fs_directory_list`?  For example, if for table "abc" we check for objects with prefix "abc", and there exists "abcdef", wouldn't that match?

**Review by ddanderson — COMMENTED** [review]
I had one question before I sign off on this.  Everything else LGTM.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:32]: Good catch. At some point I thought of that but then never went back to do that. I will fix it tomorrow.

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [reply src/tiered/tiered_handle.c:564]: Agreed.  I was just sharing an insight.

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [inline src/tiered/tiered_handle.c:44]: Do we need both `obj_name` and `obj_uri`?  It looks like you could just use one throughout this code.

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [inline src/tiered/tiered_handle.c:58]: Does the storage source guarantee that only `.wtobj` names will be returned?  Otherwise we can't control what users might put on our bucket and shouldn't ASSERT.

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [inline src/tiered/tiered_handle.c:40]: I'm concerned that a reader might to understand the issue here just from seeing the phrase "overlapping naming."  Maybe expand this to say something about distinguishing a case where other table name is a superstring or the one we're dropping?

**Review by keitharnoldsmith — APPROVED** [review]
LGTM modulo a few minor comments

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:44]: The full URI is returned by `wt_tiered_name` in `obj_uri`. I want to use just the name portion so `obj_name` strips off the leading `object:` from the URI. The file names returned by `list_directory` are just the names which is why I use `obj_name`.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:58]: Removed.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:40]: Changed.

**Comment by sueloverso** [comment]
evergreen retry

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [reply src/tiered/tiered_handle.c:45]: Ah yes.  My mistake.

File: src/tiered/tiered_handle.c
Comment (ddanderson) [inline src/tiered/tiered_handle.c:15]: I had a long thought about naming and uniqueness, that I wrote up in WT-7702.
Still, I think the approach you're using is fine for now.

**Review by ddanderson — APPROVED** [review]
LGTM.

**Comment by sueloverso** [comment]
Moving the ticket WT-7375 to POC2 and keeping this branch as a draft for when that ticket is picked up again.

File: src/schema/schema_drop.c
Comment (keitharnoldsmith) [inline src/schema/schema_drop.c:191]: Does the compiler require `WT_UNUSED` here? We used `cfg` a few lines up.

**Review by keitharnoldsmith — APPROVED** [review]
One minor question about the code, but otherwise LGTM!

File: src/schema/schema_drop.c
Comment (sueloverso) [reply src/schema/schema_drop.c:191]: Good catch. Removing.

File: src/schema/schema_drop.c
Comment (ddanderson) [inline src/schema/schema_drop.c:205]: It looks like `ret` is ignored here?

File: src/schema/schema_drop.c
Comment (ddanderson) [inline src/schema/schema_drop.c:230]: Maybe also call `__wt_meta_track_drop`?

File: src/tiered/tiered_handle.c
Comment (ddanderson) [inline src/tiered/tiered_handle.c:48]: Consider calling `bucket_fs->fs_directory_list` at this point, using `obj_name` instead of `name`.  That should give a tiny list (if any) and save a lot of processing - especially when we get to thousands of objects.

**Review by ddanderson — APPROVED** [review]
Looks good, modulo a few suggestions/questions.

File: src/schema/schema_drop.c
Comment (sueloverso) [reply src/schema/schema_drop.c:230]: I had to think about this, but I think you're right, we need that here too. Fixed.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [reply src/tiered/tiered_handle.c:48]: I am not following this. I do call `bucket_fs->fs_directory_list` on line 34. And that is the tiny list I'm comparing against. Or I think I am. Can you please let me know what I'm missing here?

Comment (ddanderson) [reply src/tiered/tiered_handle.c:48]: I think misunderstood what the function does - so disregard that comment.  Just so I understand, it looks like if we are talking about `table:foo`, then (forgetting about prefixes), we're looking for names that match `foo-0.wtobj`, `foo-1.wtobj` etc.?  So you generate `foo-1.wtobj` and list matches for "foo" and see if any of the lengths of the result match length of `foo-1.wtobj`?  But what if (after GC is implemented in the future), lower numbers have been removed and there exists only `foo-25.wtobj` through`foo-35.wtobj`?  So I think the matching has to be more sophisticated than just a length match.

Comment (sueloverso) [reply src/tiered/tiered_handle.c:48]: The length is zero-filled and fixed for a name `foo`. It will be `foo-00001.wtobj, foo-00025.wtobj, foo-00321.wtobj` etc. (There are a lot more zeroes in the real definition). I was using the length plus name `foo` so that it does not match `fo` or `food` or even a table named `foo-00001`.

Comment (ddanderson) [reply src/tiered/tiered_handle.c:48]: Right, I forgot that.  Might be worth adding a few words reminding that the object name suffixes are fixed length.

Comment (sueloverso) [reply src/tiered/tiered_handle.c:48]: Fixed.

File: src/schema/schema_drop.c
Comment (sueloverso) [reply src/schema/schema_drop.c:205]: I missed this comment the first time through. Good catch. There were two places, both now fixed.

**Pull request merged**
